### ISSUE TITLE
Cannot extract diagonal vector from sparse matrix

### ISSUE THREAD

**Issue by neumannrf** [conradsnicta/armadillo-code#36]
Since version `8.100` I am getting build error messages when trying to extract the diagonal of an `arma::SpMat<double>`.
```
error: conversion from 'arma::spdiagview<double>' to non-scalar type 'arma::vec 
{aka arma::Col<double>}' requested
   arma::vec matrix_diagonal = matrix.diag();
```

If I try to convert it to an `arma::vec`, I get the following build error message
```
error: no matching function for call to 'arma::conv_to<arma::Col<double> >::from(arma::sp
diagview<double>)'
   arma::vec matrix_diagonal = arma::conv_to<arma::vec>::from(matrix.diag());
```

Also, if I try to use `arma::diagvec()` instead, I get
```
error: no matching function for call to 'diagvec(arma::SpMat<double>&)'
   arma::vec matrix_diagonal = arma::diagvec(matrix);
```

How can I convert an `arma::spdiagview` to an `arma::vec`?

**Comment by neumannrf** [comment]
## `example.cpp`
```C++
#include <armadillo>
using namespace arma;
int main() {
  sp_mat M(2,2);
  vec V = M.diag();
  return 0;
}
```

## `g++ example.cpp -O2 -larmadillo`
```
example.cpp: In function 'int main()':
example.cpp:6:17: error: conversion from 'arma::spdiagview<double>' to non-scalar type 'arma::vec {aka arma::Col<double>}' requested
   vec V = M.diag();
           ~~~~~~^~
```

**Comment by conradsnicta** [comment]
This has slightly changed in Armadillo 8.x. To convert a sparse diagonal view to a dense column vector, 
use an explicit constructor.  For example:
    sp_mat M(2,2);
    vec V(M.diag());
The above code will work with Armadillo 8.x and 7.x.

The support sparse diagonal views in 7.x was incomplete and hence a minor API change in 8.x was required to fix the problem.

**Comment by conradsnicta** [comment]
@neumannrf - This has slightly changed in Armadillo 8.x. To convert a sparse diagonal view to a dense column vector, 
use an explicit constructor.  For example:

    sp_mat M(2,2);
    vec V(M.diag());  // instead of vec V = M.diag();

The above code will work with Armadillo 8.x and 7.x.

The support for sparse diagonal views in 7.x was incomplete and hence a minor API change in 8.x was required to fix several underlying issues.

**Comment by conradsnicta** [comment]
@eddelbuettel @binxiangni @thirdwing - to summarise: there is a minor API change in Armadillo 8.x as described above, and the required modifications to user code are tiny.

**Comment by neumannrf** [comment]
Thanks @conradsnicta for your lightning fast response and proposed solution!

Indeed the required modification are tiny and not entirely unexpected, given the major version bump. I did look for the description of this API change in the [API additions](http://arma.sourceforge.net/docs.html#api_additions) section but couldn't find anything on the newly created `arma::spdiagview` object. Maybe a **Note** or **Caveat** remark at [`.diag()`](http://arma.sourceforge.net/docs.html#diag) would suffice to ensure a smooth transition of user codes from `7.x` to `8.x`.

Lastly, but most importantly, **huge congratulations** for your work with Armadillo!

**Comment by conradsnicta** [comment]
I'll add a note to the documentation.

**Comment by conradsnicta** [comment]
@neumannrf - I've released Armadillo 8.100.1, which has faster extraction of sparse diagonals:
http://sourceforge.net/projects/arma/files/armadillo-8.100.1.tar.xz
